contracts/asset-proxy/contracts/src/bridges/KyberBridge.sol

```diff
@@ -0,0 +1,160 @@
+    ///      tokens to `to`.
+    /// @param toTokenAddress The token to give to `to`.
+    /// @param to The recipient of the bought tokens.
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param bridgeData The abi-encoeded "from" token address.
+    /// @return success The magic bytes if successful.
+    function bridgeTransferFrom(
+        address toTokenAddress,
+        address /* from */,
+        address to,
+        uint256 amount,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        TradeState memory state;
+        state.kyber = _getKyberContract();
+        state.weth = _getWETHContract();
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (state.fromTokenAddress) = abi.decode(bridgeData, (address));
+        state.fromTokenBalance = IERC20Token(state.fromTokenAddress).balanceOf(address(this));
+        if (state.fromTokenBalance == 0) {
+            // Do nothing if no input tokens.
+            return BRIDGE_SUCCESS;
+        }
+        state.minConversionRate = (10 ** 18) * amount / state.fromTokenBalance;
+        if (state.fromTokenAddress == toTokenAddress) {
+            // Just transfer the tokens if they're the same.
+            LibERC20Token.transfer(state.fromTokenAddress, to, state.fromTokenBalance);
+            return BRIDGE_SUCCESS;
+        } else if (state.fromTokenAddress != address(state.weth)) {
+            // If the input token is not WETH, grant an allowance to the exchange
+            // to spend them.
+            LibERC20Token.approve(state.fromTokenAddress, address(state.kyber), uint256(-1));
+        } else {
+            // If the input token is WETH, unwrap it and attach it to the call.
+            state.fromTokenAddress = KYBER_ETH_ADDRESS;
+            state.payableAmount = state.fromTokenBalance;
+            state.weth.withdraw(state.fromTokenBalance);
+        }
+
+        // Try to sell all of this contract's input token balance.
+        uint256 boughtAmount = state.kyber.trade.value(state.payableAmount)(
+            // Input token.
+            state.fromTokenAddress,
+            // Sell amount.
+            state.fromTokenBalance,
+            // Output token.
+            toTokenAddress == address(state.weth) ?
+                KYBER_ETH_ADDRESS :
+                toTokenAddress,
+            // Transfer to this contract if converting to ETH, otherwise
+            // transfer directly to the recipient.
+            toTokenAddress == address(state.weth) ?
+                address(uint160(address(this))) :
+                address(uint160(to)),
+            // Buy as much as possible.
+            uint256(-1),
+            // Minimum conversion rate.
+            state.minConversionRate,
+            // No affiliate address.
+            address(0)
+        );
+        // Wrap ETH output and transfer to recipient.
+        if (toTokenAddress == address(state.weth)) {
```

Comment:
So Kyber will send this contract ETH instead of WETH, right? We'll need a fallback function to accept it in that case.

Reply:
Oh woops! I totally forgot! This actually explains a confusing revert I kept getting while writing the tests. :upside_down_face: